Allow us to run tests from external XBlock repositories

pmitros opened the pull request:
This is the next step in the XBlocks test cleanup. This allows us to run tests included in XBlock repos with exported entrypoints. The whole PR is less than a screenful of code.

For an example of a working test in such a repo, I made a branch of the RateXBlock. Please see https://github.com/pmitros/RateXBlock/tree/pmitros/ux-revamp and specifically https://github.com/pmitros/RateXBlock/blob/pmitros/ux-revamp/setup.py and https://github.com/pmitros/RateXBlock/blob/pmitros/ux-revamp/ratetests/test_rate.py

As a next step, I will add a list of tests required to be there. I would like the system, in a case like this one, to also verify RateXBlock is installed. I will probably do that together with the RateXBlock PR, since right now, there are no external tests I want to guarantee.

@cpennington @cahrens 

(Also, @nedbat and @ormsbee if either of you have time, interest, or feedback)

nedbat commented on openedx/tests/xblock_integration/test_external_xblocks.py:
```diff
@@ -0,0 +1,37 @@
+
+for entrypoint in pkg_resources.iter_entry_points(group="xblock.test.v0"):
+    plugin = entrypoint.load()
+    classname = str(entrypoint).split(":")[-1]
```
At this point, `plugin` is the class itself.  No need to parse and check for validity.  Use `classname = p.__name__`.

pmitros replied:
Thank you. I will change to plugin.__name__.

I would like to keep the validity checks, however. For better or worse, the Python test frameworks do rely on naming to figure out which things are tests, and I'd like to enforce such conventions here. Even as a sanity check, I want to catch errors like: 

* Someone makes an entrypoint that points to a non-test class. E.g. they cut-and-paste in setup.py, and fail to update one field, and point to the XBlock itself, or any of a range of similar errors. 
* There are two versions of the same XBlock installed, or two XBlocks have a name collision.

It's also good security practice. Without such checks, we're giving XBlock authors freedom to inject anything into globals() here otherwise. I agree that this doesn't raise any security issues /right now/, but checks like this make it much less likely we run into security issues as the code evolves in the future.

nedbat commented on openedx/tests/xblock_integration/test_external_xblocks.py:
```diff
@@ -0,0 +1,52 @@
+    if not classname.startswith("Test"):
+        raise InvalidTestName("Test class should start with 'Test': " + classname)
+    if not classname.replace("_", "").isalnum():
+        raise InvalidTestName("Python variables should be letters, numbers, and underscores: " + classname)
```
I'm fine with the duplicate check (actually, the idea that two disparate XBlocks have to be careful not to clobber each other's tests is a concern).  I'm fine with the "Test" check.  But checking for a valid identifier is unnecessary, since an invalid identifier wouldn't be possible without malicious tricks anyway.

pmitros replied:
I generally prefer to be a little overly conservative, but if you insist, I'll remove the extra check. I think it would minimize the impact of possible malicious tricks, but in this case, there is no immediate increase in security exposure, so I can kill it. It's just two lines of code, though. Your call.

I share the concern about blocks clobbering each others' tests, but I figure we can address that when we get to it. Right now, I have exactly two test cases (`RateXBlock` and `DoneXBlock`). It wouldn't be hard to rename the tests as they come in and similar, but I would be a little concerned about failing to catch corner cases:

* Introspection/dynamic code which refers to things by name.
* Absolute imports with overlapping names
* Same test class name vs. same test+package name vs. etc.
* Overlaps in identifiers in OLX
* Etc. 

There are a lot of these corner cases here, but without generating them, and making a lot of test cases, I'd have little confidence I'd address them correctly, and even so, I wouldn't be sure I thought of all them. For now, we have 50 XBlocks in the community, which is a little bit unwieldy for tests, but issues like this can be managed on a case-by-case basis.

Note that a lot of these already come up in XBlocks. If I have two blocks which both call themselves `<video>`, the OLX parsing becomes undefined.

pmitros replied:
(come up as hypothetical, that is; I'm not aware of an actual case here)

cpennington replied:
This might be overly magical, but we could actually disambiguate the tests in this file, by generating subclasses with our own names when the names conflict.

cpennington replied:
Oh, and now that I read your comment, I see you considered that already. Carry on.

cpennington commented on openedx/tests/xblock_integration/test_external_xblocks.py:
```diff
@@ -0,0 +1,52 @@
+    print "This XBlock testing framework is in ALPHA. Please read the"
+    print "disclaimers in before using -- it may disappear at any point. When"
+    print "it moves into production, these print statements will disappear,"
+    print "and so might your tests."
```
Nit: use `textwrap.dedent`?

cpennington commented:
:+1:

pmitros commented:
Thanks Cale!

* I switched to dedent.
* I added comment pointing to this discussion for renaming in the future.
* I fixed pep8/pylint.
* I squashed commits.
* I cleaned up a few more comments.

I am merging as soon as tests pass. @nedbat (or anyone else) If you feel strongly about removing the `isalnum` test, or anything else, let me know soon, and I'll kill it before merge.

pmitros commented:
jenkins run lettuce

nedbat commented:
I'm curious where these printed disclaimers appear?  I ran a test from the RateXBlock, and the test ran, but no disclaimer appeared.  Stdout is typically buffered during tests, and only shown on test failures, but this message isn't happening within the scope of a test.

pmitros commented:
I do not know if they appear anywhere visible to developers. I could send to std err if you wanted. That's buffered as well, but in case of failure, I believe it prints out, mixed with a billion other lines send to std err. 

It's a little bit of a trade-off. I wanted this more prominent than a code comment, which never get read. On the other hand, we don't want to spam all developers all the time, since most aren't doing XBlock work. I thought this was a sort of okay trade-off, since this makes it very prominent if anyone ever looks at the source file (which is still largely prerequisite to building tests this way, so long as it is in development).

In the end, my plan is to replace this with: 

    if len(pkg_resources.iter_entry_points(group="xblock.test.v0")):
        raise DeprecatedException()
    for entrypoint in pkg_resources.iter_entry_points(group="xblock.test.v1"):
        ....

Or possibly the same, but only raising the exception if an entry point is in v0 but not in v1 (so developers can supply both during the transition). 

There is also a chance we may not need to deprecate, but I'd need @benpatterson's feedback on how we'll do acceptance tests to know.

pmitros commented:
Ned and I spoke in person. `print` statement removed at his insistence.
